cmd/blsync/main.go

```diff
@@ -0,0 +1,243 @@
+		FinalizedBlockHash: finalizedHash,
+	}
+	ctx, cancel := context.WithTimeout(context.Background(), time.Second*5)
+	err := client.CallContext(ctx, &resp, "engine_forkchoiceUpdatedV1", update, nil)
+	cancel()
+	return resp.PayloadStatus.Status, err
+}
+
+type execSyncer struct {
+	api           *api.BeaconLightApi
+	client        *rpc.Client
+	execRootCache *lru.Cache[common.Hash, common.Hash] // beacon block root -> execution block root
+}
+
+var statePaths = []string{
+	"[\"finalizedCheckpoint\",\"root\"]",
+	"[\"latestExecutionPayloadHeader\",\"blockHash\"]",
+}
+
+func (e *execSyncer) newHead(head types.Header) {
+	log.Info("Received new beacon head", "slot", head.Slot, "blockRoot", head.Hash())
+	if e.client == nil {
+		return
+	}
+	var (
+		blocks         []*ctypes.Block
+		finalizedRoots []common.Hash
+		parentFound    bool
+		maxBlockFetch  = 4
+	)
+	for {
+		proof, err := e.api.GetStateProof(head.StateRoot, statePaths, stateProofFormat)
+		if err != nil {
+			log.Error("Error fetching state proof from beacon API", "error", err)
+			break
+		}
+		var (
+			execBlockRoot       = common.Hash(proof.Values[execBlockIndex])
+			finalizedBeaconRoot = common.Hash(proof.Values[finalizedBlockIndex])
+			beaconRoot          = head.Hash()
+		)
+		block, err := e.api.GetExecutionPayload(beaconRoot, execBlockRoot)
+		if err != nil {
+			log.Error("Error fetching execution payload from beacon API", "error", err)
+			break
+		}
+		blocks = append(blocks, block)
+		finalizedRoots = append(finalizedRoots, finalizedBeaconRoot)
+		e.execRootCache.Add(beaconRoot, execBlockRoot)
+
+		if _, ok := e.execRootCache.Get(head.ParentRoot); ok {
+			parentFound = true
+			break
+		}
+		maxBlockFetch--
+		if maxBlockFetch == 0 {
+			break
+		}
+
+		if head, err = e.api.GetHeader(head.ParentRoot); err != nil {
+			log.Error("Error fetching header from beacon API", "error", err)
+			break
+		}
+	}
+	if blocks == nil {
+		return
+	}
+	if !parentFound && maxBlockFetch == 0 {
+		blocks = blocks[:1]
+		finalizedRoots = finalizedRoots[:1]
+		// iterate further back as far as possible (or max 256 steps) just to collect beacon -> exec root associations
+		maxFinalizedFetch := 256
+		for {
+			proof, err := e.api.GetStateProof(head.StateRoot, statePaths, stateProofFormat)
+			if err != nil {
+				// exit silently because we expect running into an error
+				break
+			}
+			execBlockRoot := common.Hash(proof.Values[execBlockIndex])
+			beaconRoot := head.Hash()
+			e.execRootCache.Add(beaconRoot, execBlockRoot)
+			if beaconRoot == finalizedRoots[0] {
+				break // this is what we were looking for, earlier associations won't be needed
+			}
+			if _, ok := e.execRootCache.Get(head.ParentRoot); ok {
+				break
+			}
+			maxFinalizedFetch--
+			if maxFinalizedFetch == 0 {
+				break
+			}
+			if head, err = e.api.GetHeader(head.ParentRoot); err != nil {
+				break
+			}
+		}
+	}
+	for i := len(blocks) - 1; i >= 0; i-- {
+		blockRoot := blocks[i].Hash()
```

Comment:
On a completely separate note... `blocks[i]` is `*ctypes.Block`. If the `ctype.Block` method `Hash()` returns the _stateRoot_ as opposed to the blockhash, then I think that is bad API, and not what I would expect from somthing called "block".

Reply:
I don't follow. State root was never mentioned here, these are all block roots. `ctypes` is `core/types` so we are talking about good old execution blocks here and their `Hash()` definitely returns the block root. The beacon header's `Hash()` function does the same of course, state root is accessible as a field in both cases.